egs/wsj/s5/steps/nnet3/fvector/generate_fixed_length_range.py

```diff
@@ -0,0 +1,191 @@
+# If the noise file is shorter than T. We select the whole noise.
+
+# The <snr> control the rate of signal and noise. In the other word, scale the amplitude of noise.
+# The snr will be randomly selected form the range (max-snr, min-snr).
+
+from __future__ import print_function
+import re, os, argparse, sys, math, warnings, random
+
+parser = argparse.ArgumentParser(description="Generate a noise range files which contains "
+                                 "N lines corresponding to the number of kinds for each original wav. "
+                                 "The file created by this python code will be supplied to fixed-length add "
+                                 "additive noise program.",
+                                 epilog="Called by steps/nnet3/fvector/add_noise.sh")
+parser.add_argument("--num-kind-range", type=int, default=4,
+                    help="the number of expected addnoise kinds")
+parser.add_argument("--min-additive-noise-len", type=float, default=2.0,
+                    help="the minimum duration/length of each noise file")
+parser.add_argument("--min-snr", type=int, default=0,
+                    help="the minimum Signal-to-Noise Rate, the default=0")
+parser.add_argument("--max-snr", type=int, default=-10,
+                    help="the maximum Signal-to-Noise Rate, the default=-10")
+parser.add_argument("--seed", type=int, default=-1,
+                    help="Seed for random number generator")
+
+# now the positional arguments
+parser.add_argument("wav2dur",
+                    help="wav2dur file of the original wav to be used as input (format is: "
+                    "<utterance-id> <duration>")
+parser.add_argument("noise2dur",
+                    help="noise2dur file of the noise wav to be used as input (format is: "
+                    "<utterance-id> <duration>")
+parser.add_argument("range_dir",
+                    help="Name of ranges directory, exp/fxvector/ranges")
+
+print(' '.join(sys.argv))
+
+args = parser.parse_args()
+
+## Check arguments
+if args.min_snr < args.max_snr:
+    sys.exit("For SNR, the less numerical value is, the larger noise is. So --min-snr bigger "
+             "than --max-snr in numerical value.")
+
+random.seed(args.seed)
+
+# deal with the original wav utt2dur
+# the information was stored in wav_ids[], wav_lengths[] and wav_num_utts
+f = open(args.wav2dur, "r")
+if f is None:
+    sys.exit("Error opening wav2dur file " + str(args.wav2dur))
+wav_ids = []
+wav_lengths = []
+for line in f:
+    a = line.split()
+    if len(a) != 2:
+        sys.exit("Bad line in wav2dur file " + line)
+    if float(a[1]) < args.min_additive_noise_len:
+	sys.exit("ERROR: The wav length is shorter than --min-additive-noise-len")
+    wav_ids.append(a[0])
+    wav_lengths.append(float(a[1]))
+f.close()
+
+wav_num_utts = len(wav_ids)
+
+# deal with the noise wav utt2dur
+# remove the noise whose length < --min-additive-noise-len
+num_error = 0
+num_done = 0
+f = open(args.noise2dur, "r")
```

Comment:
It is better to add separate function ass wave2dur with output as wav_id and wav_len and use it for both noise2dur and wav2dur.